R/plotting.R

```diff
@@ -18,6 +18,47 @@
 # suppressPackageStartupMessages(library(d3r))
 # suppressPackageStartupMessages(library(viridis))
 
+########################
+## Internal Functions ##
+########################
+#' 
+#' 
+.LevelReduction <- function(lin, level) {   
```

Comment:
Can we have boundary and Error Handling:

Where guards are added in .LevelReduction to handle cases where the lineage string does not contain enough ">" delimiters, returning the original lineage if the specified level is out of bounds. This change enhances error resilience. An example is thus:

 .LevelReduction <- function(lin, level) {
    # Check if '>' delimiter is present in the lineage
    gt_loc <- str_locate_all(lin, ">")[[1]] 
    
    # Calculate the number of levels available in the lineage
    available_levels <- length(gt_loc) / 2  # Since `str_locate_all` returns a matrix
    
    # Guard against out-of-bounds level requests
    if (level > available_levels || level < 1) {
        # Not enough '>' in lineage or invalid level requested
        return(lin)
    } else {
        # Retrieve the location of the requested '>' delimiter
        gt_loc <- gt_loc[level, ][1] %>% as.numeric()
        # Extract lineage up to the specified level
        lin <- substring(lin, first = 0, last = (gt_loc - 1))
        return(lin)
    }
}

Explanation:
str_locate_all(lin, ">")[[1]]: Finds all occurrences of > in the lineage string.
Boundary Guard: Checks if the requested level is within bounds:
If level is greater than available_levels (i.e., there aren’t enough > delimiters) or if level is less than 1, it returns the original lineage.
Lineage Extraction: If level is within bounds, the function extracts the lineage up to the specified level.
This boundary check ensures that .LevelReduction gracefully handles cases where the requested level is out of bounds, thereby improving error resilience.

Reply:
@Joiejoie1 Have a look at this implementation. A little more concise but now checks if the specified level exists within the lineage and then calculating the total available levels. it would return the full lineage string if the requested level exceeds the total levels or if the input is invalid (i.e. negative or zero)

```
.LevelReduction <- function(lin, level) {
    gt_loc <- str_locate_all(lin, ">")[[1]]
    available_levels <- length(gt_loc) / 2  # Since `str_locate_all` returns a matrix
    
    # Guard against out-of-bounds level requests
    if (level > available_levels || level < 1) {
        return(lin)
    } else {
        gt_loc <- gt_loc[level, ][1] %>% as.numeric()
        lin <- substring(lin, first = 0, last = (gt_loc - 1))
        return(lin)
    }
}
```

Reply:
> @Joiejoie1 Have a look at this implementation. A little more concise but now checks if the specified level exists within the lineage and then calculating the total available levels. it would return the full lineage string if the requested level exceeds the total levels or if the input is invalid (i.e. negative or zero)
> 
> ```
> .LevelReduction <- function(lin, level) {
>     gt_loc <- str_locate_all(lin, ">")[[1]]
>     available_levels <- length(gt_loc) / 2  # Since "str_locate_all" returns a matrix
>     
>     # Guard against out-of-bounds level requests
>     if (level > available_levels || level < 1) {
>         return(lin)
>     } else {
>         gt_loc <- gt_loc[level, ][1] %>% as.numeric()
>         lin <- substring(lin, first = 0, last = (gt_loc - 1))
>         return(lin)
>     }
> }
> ```

Thank you, @teddyCodex, for sharing this! I appreciate the streamlined approach, especially the additional checks to handle cases where the specified level exceeds available levels or is invalid. Using str_locate_all to check the number of levels is a smart way to keep the function concise and error-proof. I’ll go through it and test how it performs across cases with different lineage structures to confirm everything works smoothly.

Let me know if there’s anything specific you’d like me to focus on in the review.